tests/csapi/room_upgrade_test.go

```diff
@@ -0,0 +1,309 @@
+
+	alice := deployment.Client(t, "hs1", "@alice:hs1")
+
+	roomID := alice.CreateRoom(t, map[string]interface{}{
+		"preset": "public_chat",
+	})
+
+	_, tombstoneSinceToken := alice.MustSync(t, client.SyncReq{TimeoutMillis: "0"})
+	_, sinceToken := alice.MustSync(t, client.SyncReq{TimeoutMillis: "0"})
+
+	resp := alice.MustDoFunc(
+		t,
+		"POST",
+		[]string{"_matrix", "client", "v3", "rooms", roomID, "upgrade"},
+		client.WithJSONBody(t, map[string]string{
+			"new_version": "6",
+		}),
+	)
+
+	upgradeJson := gjson.ParseBytes(must.ParseJSON(t, resp.Body))
+	must.MatchGJSON(
+		t,
+		upgradeJson,
+		match.JSONKeyPresent("replacement_room"),
+		match.JSONKeyTypeEqual("replacement_room", gjson.String),
+	)
+	replacementRoom := upgradeJson.Get("replacement_room").Str
+
+	if roomID == replacementRoom {
+		t.Fatalf("Rooms are the same (%s)", roomID)
+	}
+
+	// Check for and fetch the tombstone event ID
+	tombstoneEventId := ""
+	alice.MustSyncUntil(t, client.SyncReq{Since: tombstoneSinceToken}, client.SyncTimelineHas(roomID, func(ev gjson.Result) bool {
+		if ev.Get("type").Str == "m.room.tombstone" && ev.Get("state_key").Exists() {
+			repRoomRes := ev.Get("content.replacement_room")
+
+			if repRoomRes.Exists() {
+				tombstoneEventId = ev.Get("event_id").Str
+
+				if repRoomRes.Str != replacementRoom {
+					t.Errorf("tombstone event did not match replacement room, got %s, want %s", repRoomRes.Str, replacementRoom)
+				}
+
+				return true
+			}
+		}
+
+		return false
+	}))
+
+	// Convenience method to check for state event
+	stateExists := func(evType string) func(gjson.Result) bool {
+		return func(ev gjson.Result) bool {
+			return ev.Get("type").Str == evType && ev.Get("state_key").Exists()
+		}
+	}
+
+	// Note: we cannot check for the following state here, as it's only prescribed as an "implementation detail" by the spec:
+	// - m.room.guest_access
+	// - m.room.history_visibility
+	// - m.room.join_rules
+	alice.MustSyncUntil(t, client.SyncReq{Since: sinceToken}, client.SyncJoinedTo(alice.UserID, replacementRoom),
+		client.SyncTimelineHas(replacementRoom, stateExists("m.room.create")),
+		client.SyncTimelineHas(replacementRoom, stateExists("m.room.member")),
+		client.SyncTimelineHas(replacementRoom, stateExists("m.room.power_levels")),
+
+		client.SyncTimelineHas(replacementRoom, func(ev gjson.Result) bool {
+			if ev.Get("type").Str == "m.room.create" && ev.Get("state_key").Exists() {
+				predecessor := ev.Get("content.predecessor")
+
+				if predecessor.Get("room_id").Exists() && predecessor.Get("event_id").Exists() {
+					if predecessor.Get("room_id").Str != roomID {
+						t.Errorf("predecessor room id did not match: got %s, want %s", predecessor.Get("room_id").Str, roomID)
+					}
+
+					if predecessor.Get("event_id").Str != tombstoneEventId {
+						t.Errorf("predecessor tombstone event id did not match: got %s, want %s", predecessor.Get("event_id").Str, tombstoneEventId)
+					}
+
+					return true
+				}
+			}
```

Comment:
The old sytest used to assert that the upgraded room has the requested room version. I don't think this is done here(?)